planner/core/logical_plan_builder.go

```diff
@@ -856,6 +860,152 @@ func (b *PlanBuilder) buildJoin(ctx context.Context, joinNode *ast.Join) (Logica
+	b.handleHelper.mergeAndPush(handleMap1, handleMap2)
+
+	leftJoinPlan := LogicalJoin{StraightJoin: joinNode.StraightJoin || b.inStraightJoin}.Init(b.ctx, b.getSelectOffset())
+	leftJoinPlan.SetChildren(lChild1, rChild1)
+	leftJoinPlan.SetSchema(expression.MergeSchema(lChild1.Schema(), rChild1.Schema()))
+	leftJoinPlan.names = make([]*types.FieldName, lChild1.Schema().Len()+rChild1.Schema().Len())
+	copy(leftJoinPlan.names, lChild1.OutputNames())
+	copy(leftJoinPlan.names[lChild1.Schema().Len():], rChild1.OutputNames())
+
+	leftJoinPlan.JoinType = LeftOuterJoin
+
+	// Merge sub-plan's fullSchema into this join plan.
+	// Please read the comment of LogicalJoin.fullSchema for the details.
+	var (
+		lFullSchema, rFullSchema *expression.Schema
+		lFullNames, rFullNames   types.NameSlice
+	)
+	if left, ok := lChild1.(*LogicalJoin); ok && left.fullSchema != nil {
+		lFullSchema = left.fullSchema
+		lFullNames = left.fullNames
+	} else {
+		lFullSchema = lChild1.Schema()
+		lFullNames = lChild1.OutputNames()
+	}
+	if right, ok := rChild1.(*LogicalJoin); ok && right.fullSchema != nil {
+		rFullSchema = right.fullSchema
+		rFullNames = right.fullNames
+	} else {
+		rFullSchema = rChild1.Schema()
+		rFullNames = rChild1.OutputNames()
+	}
+	if joinNode.Tp == ast.RightJoin {
+		// Make sure lFullSchema means outer full schema and rFullSchema means inner full schema.
+		lFullSchema, rFullSchema = rFullSchema, lFullSchema
+		lFullNames, rFullNames = rFullNames, lFullNames
+	}
+	leftJoinPlan.fullSchema = expression.MergeSchema(lFullSchema, rFullSchema)
+
+	// Merge sub-plan's fullNames into this join plan, similar to the fullSchema logic above.
+	leftJoinPlan.fullNames = make([]*types.FieldName, 0, len(lFullNames)+len(rFullNames))
+	for _, lName := range lFullNames {
+		name := *lName
+		leftJoinPlan.fullNames = append(leftJoinPlan.fullNames, &name)
+	}
+	for _, rName := range rFullNames {
+		name := *rName
+		leftJoinPlan.fullNames = append(leftJoinPlan.fullNames, &name)
+	}
+
+	// Clear NotNull flag for the inner side schema if it's an outer join.
+	if joinNode.Tp == ast.LeftJoin || joinNode.Tp == ast.RightJoin {
+		resetNotNullFlag(leftJoinPlan.fullSchema, 0, leftJoinPlan.fullSchema.Len())
+	}
+
+	if joinNode.On != nil {
+		b.curClause = onClause
+		onExpr, newPlan, err := b.rewrite(ctx, joinNode.On.Expr, leftJoinPlan, nil, false)
+		if err != nil {
+			return nil, err
+		}
+		if newPlan != leftJoinPlan {
+			return nil, errors.New("ON condition doesn't support subqueries yet")
+		}
+		onCondition := expression.SplitCNFItems(onExpr)
+		leftJoinPlan.AttachOnConds(onCondition)
+	}
+
+	// build the part of anti join.
+	lChild2, err := b.buildResultSetNode(ctx, joinNode.Left, false)
+	if err != nil {
+		return nil, err
+	}
+	rChild2, err := b.buildResultSetNode(ctx, joinNode.Right, false)
+	if err != nil {
+		return nil, err
+	}
+	handleMap1 = b.handleHelper.popMap()
+	handleMap2 = b.handleHelper.popMap()
+	b.handleHelper.mergeAndPush(handleMap1, handleMap2)
```

Comment:
Does the order of `handleMap1` and `handleMap2` matter here?

Reply:
No